### ISSUE TITLE
Error calling jty with a simple jsonnet file

### ISSUE THREAD

**Issue by BenjaminHerbert** [mark-rushakoff/jty#1]
I am encountering an error that while trying out jty with the following jsonnet file taken from here: https://jsonnet.org/learning/getting_started.html

```
> cat fuu.jsonnet 
{
   "person1": {
      "name": "Alice",
      "welcome": "Hello Alice!"
   },
   "person2": {
      "name": "Bob",
      "welcome": "Hello Bob!"
   }
}
```

The error message is:
```
> jty   fuu.jsonnet fuu.yaml
failed to evaluate jsonnet at fuu.jsonnet: RUNTIME ERROR: stream mode: top-level object was a object, should be an array whose elements hold the JSON for each document in the stream.
	During manifestation	

encountered errors during processing; failing
```

Could you shed some light what I am doing or understanding wrong? :confused:

**Comment by mark-rushakoff** [comment]
`jty` expects .jsonnet files to emit arrays, not single objects. So if you adjust fuu.jsonnet like this:

```jsonnet
[
  {
    person1: {
      name: 'Alice',
      welcome: 'Hello Alice!',
    },
    person2: {
      name: 'Bob',
      welcome: 'Hello Bob!',
    },
  },
]
```

Then you will be able to generate this fuu.yaml:

```yaml
---
person1:
    name: Alice
    welcome: Hello Alice!
person2:
    name: Bob
    welcome: Hello Bob!
...
```

In our jsonnet use cases at work, all of our .jsonnet files emit multiple objects, so the single-object output actually had not come up.

Glancing through the jsonnet code called in that path, I'm not sure how simple it will be to allow for a single object rather than an array.

**Comment by BenjaminHerbert** [comment]
Thanks for your hint. It was just a very simple example to illustrate the problem.

I am still thinking how to apply this to my files without changing their structure too much.

I am able to run jty when I "wrap an array" around my original file like this:
```
[ 
 (import originial-file.jsonnet)
]
```
So perhaps I could "wrap an array around each file", if there is none... :thinking: 

Could you give me a hint as to what is the crucial path to prevent (the VM?) re-evaluating the import files?
Is it this `p.vm.EvaluateSnippetStream(req.InPath, req.JsonnetContent)`  from [jty/processor.go](https://github.com/mark-rushakoff/jty/blob/2635698512dfe2007bb4ef5e6138f2b2310eab62/pkg/jty/processor.go#L147)? ?

**Comment by mark-rushakoff** [comment]
Correct, the crucial path here is the single instance of `(*Processor).vm`.

It's been a while, but IIRC, the Jsonnet VM caches everything about a single file. `jty` has some clever code to ingest and output files concurrently, but to effectively call the Jsonnet VM in a single thread; since it reuses a single VM, we don't repeatedly parse commonly imported files. That is, if you have

```jsonnet
// a.jsonnet
local foo = import './foo.libsonnet'

foo.a
```

and

```jsonnet
// b.jsonnet
local foo = import './foo.libsonnet'

foo.b
```

then reusing the one VM, our process only opens the `foo.jsonnet` file once and only parses its content once.

If all of your Jsonnet files output a single object rather than an array of objects, you could change `p.vm.EvaluateSnippetStream` in your link to `p.vm.EvaluateSnippet`, and then wrap the returned json string in a slice.

I could imagine jty flags like `--single` and `--multi` which set the mode of the Jsonnet parser, like:

```
jty --single a.jsonnet a.yml b.jsonnet b.yml --multi two.jsonnet two.yml
```

which would treat `a` and `b` as containing single objects, but `two` is expected to contain an array.

Or, perhaps jty could always use `EvaluateSnippet`, and it could automatically wrap single objects in an array, and leave arrays unmodified.

The flags feel a bit too manual, but the inference seems a little risky to me too. I don't have a grasp on whether there are projects in the wild that have a mix of single- and multi-object Jsonnet files.
